File: sdk/src/events.ts
// ============================================================================
// Event log decoder
// ============================================================================
// Mirrors programs/protocol/src/events.rs. The Pinocchio-era program emits
// each event as one log line of the form
//
//   Program log: EVT:<base64>
//
// where the base64 payload is `[1-byte tag][Borsh body]`. We strip the
// `Program log: ` runtime prefix, the `EVT:` marker, decode base64, peel
// the tag, and Borsh-deserialize the remainder into the matching struct.

import type { Connection, PublicKey } from "@solana/web3.js";
import BN from "bn.js";

import { Reader } from "./borsh.js";
import { PROGRAM_ID, type AnchorProvider } from "./program.js";

// ----------------------------------------------------------------------------
// Event payload shapes (1:1 with programs/protocol/src/events.rs)
// ----------------------------------------------------------------------------

export interface PoolInitializedData {
  pool: PublicKey;
  admin: PublicKey;
  oracleSigner: PublicKey;
  baseMint: PublicKey;
  quoteMint: PublicKey;
  initialFairValue: BN;
  initialSpreadBps: number;
  initialModeTtl: number;
  slot: BN;
}

export interface OracleUpdatedData {
  pool: PublicKey;
  oracleSigner: PublicKey;
  newFairValue: BN;
  newSpreadBps: number;
  newNonce: BN;
  newTtl: number;
  slot: BN;
}

export interface SwapExecutedData {
  pool: PublicKey;
  user: PublicKey;
  /** 0 = Buy, 1 = Sell */
  direction: number;
  /** 0 = curve path, 1 = RFQ path */
  mode: number;
  inputAmount: BN;
  outputAmount: BN;
  executionPrice: BN;
  quoteNonce: BN;
  slot: BN;
}

export interface PoolPausedChangedData {
  pool: PublicKey;
  admin: PublicKey;
  paused: boolean;
  slot: BN;
}

export interface OracleSignerRotatedData {
  pool: PublicKey;
  admin: PublicKey;
  previousSigner: PublicKey;
  newSigner: PublicKey;
  slot: BN;
}

export interface AdminRotatedData {
  pool: PublicKey;
  previousAdmin: PublicKey;
  newAdmin: PublicKey;
  slot: BN;
}

export interface InventoryWithdrawnData {
  pool: PublicKey;
  admin: PublicKey;
  baseAmount: BN;
  quoteAmount: BN;
  slot: BN;
}

export interface QuoteMarkerClosedData {
  pool: PublicKey;
  closer: PublicKey;
  nonce: BN;
  expirySlot: BN;
  slot: BN;
}

export interface AdminProposalCreatedData {
  pool: PublicKey;
  proposedBy: PublicKey;
  newAdmin: PublicKey;
  slot: BN;
}

export interface AdminProposalCancelledData {
  pool: PublicKey;
  admin: PublicKey;
  cancelledNewAdmin: PublicKey;
  slot: BN;
}

export interface ProtocolEventDataMap {
  PoolInitialized: PoolInitializedData;
  OracleUpdated: OracleUpdatedData;
  SwapExecuted: SwapExecutedData;
  PoolPausedChanged: PoolPausedChangedData;
  OracleSignerRotated: OracleSignerRotatedData;
  AdminRotated: AdminRotatedData;
  InventoryWithdrawn: InventoryWithdrawnData;
  QuoteMarkerClosed: QuoteMarkerClosedData;
  AdminProposalCreated: AdminProposalCreatedData;
  AdminProposalCancelled: AdminProposalCancelledData;
}

export type ProtocolEventName = keyof ProtocolEventDataMap;

export type DecodedEvent = {
  [K in ProtocolEventName]: { name: K; data: ProtocolEventDataMap[K] };
}[ProtocolEventName];

// ----------------------------------------------------------------------------
// Base64 decoder (RFC 4648 alphabet, accepts padded and unpadded inputs)
// ----------------------------------------------------------------------------

const BASE64_TABLE: Int8Array = (() => {
  const t = new Int8Array(256).fill(-1);
  const a = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  for (let i = 0; i < a.length; i++) t[a.charCodeAt(i)] = i;
  // also accept URL-safe variants
  t["-".charCodeAt(0)] = 62;
  t["_".charCodeAt(0)] = 63;
  return t;
})();

function base64Decode(s: string): Uint8Array | null {
  // Drop padding.
  while (s.endsWith("=")) s = s.slice(0, -1);
  const out = new Uint8Array(Math.floor((s.length * 3) / 4));
  let oi = 0;
  let buf = 0;
  let bits = 0;
  for (let i = 0; i < s.length; i++) {
    const v = BASE64_TABLE[s.charCodeAt(i)];
    if (v < 0) return null;
    buf = (buf << 6) | v;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out[oi++] = (buf >> bits) & 0xff;
    }
  }
  return out.subarray(0, oi);
}

// ----------------------------------------------------------------------------
// Per-event decoders
// ----------------------------------------------------------------------------

const TAG_POOL_INITIALIZED = 0x01;
const TAG_ORACLE_UPDATED = 0x02;
const TAG_SWAP_EXECUTED = 0x03;
const TAG_POOL_PAUSED_CHANGED = 0x04;
const TAG_ORACLE_SIGNER_ROTATED = 0x05;
const TAG_ADMIN_ROTATED = 0x06;
const TAG_INVENTORY_WITHDRAWN = 0x07;
const TAG_QUOTE_MARKER_CLOSED = 0x08;
const TAG_ADMIN_PROPOSAL_CREATED = 0x09;
const TAG_ADMIN_PROPOSAL_CANCELLED = 0x0a;

function decodeBody(tag: number, body: Uint8Array): DecodedEvent | null {
  const r = new Reader(body);
  try {
    switch (tag) {
      case TAG_POOL_INITIALIZED:
        return {
          name: "PoolInitialized",
          data: {
            pool: r.pubkey(),
            admin: r.pubkey(),
            oracleSigner: r.pubkey(),
            baseMint: r.pubkey(),
            quoteMint: r.pubkey(),
            initialFairValue: r.u64(),
            initialSpreadBps: r.u16(),
            initialModeTtl: r.u8(),
            slot: r.u64(),
          },
        };
      case TAG_ORACLE_UPDATED:
        return {
          name: "OracleUpdated",
          data: {
            pool: r.pubkey(),
            oracleSigner: r.pubkey(),
            newFairValue: r.u64(),
            newSpreadBps: r.u16(),
            newNonce: r.u64(),
            newTtl: r.u8(),
            slot: r.u64(),
          },
        };
      case TAG_SWAP_EXECUTED:
        return {
          name: "SwapExecuted",
          data: {
            pool: r.pubkey(),
            user: r.pubkey(),
            direction: r.u8(),
            mode: r.u8(),
            inputAmount: r.u64(),
            outputAmount: r.u64(),
            executionPrice: r.u64(),
            quoteNonce: r.u64(),
            slot: r.u64(),
          },
        };
      case TAG_POOL_PAUSED_CHANGED:
        return {
          name: "PoolPausedChanged",
          data: {
            pool: r.pubkey(),
            admin: r.pubkey(),
            paused: r.u8() !== 0,
            slot: r.u64(),
          },
        };
      case TAG_ORACLE_SIGNER_ROTATED:
        return {
          name: "OracleSignerRotated",
          data: {
            pool: r.pubkey(),
            admin: r.pubkey(),
            previousSigner: r.pubkey(),
            newSigner: r.pubkey(),
            slot: r.u64(),
          },
        };
      case TAG_ADMIN_ROTATED:
        return {
          name: "AdminRotated",
          data: {
            pool: r.pubkey(),
            previousAdmin: r.pubkey(),
            newAdmin: r.pubkey(),
            slot: r.u64(),
          },
        };
      case TAG_INVENTORY_WITHDRAWN:
        return {
          name: "InventoryWithdrawn",
          data: {
            pool: r.pubkey(),
            admin: r.pubkey(),
            baseAmount: r.u64(),
            quoteAmount: r.u64(),
            slot: r.u64(),
          },
        };
      case TAG_QUOTE_MARKER_CLOSED:
        return {
          name: "QuoteMarkerClosed",
          data: {
            pool: r.pubkey(),
            closer: r.pubkey(),
            nonce: r.u64(),
            expirySlot: r.u64(),
            slot: r.u64(),
          },
        };
      case TAG_ADMIN_PROPOSAL_CREATED:
        return {
          name: "AdminProposalCreated",
          data: {
            pool: r.pubkey(),
            proposedBy: r.pubkey(),
            newAdmin: r.pubkey(),
            slot: r.u64(),
          },
        };
      case TAG_ADMIN_PROPOSAL_CANCELLED:
        return {
          name: "AdminProposalCancelled",
          data: {
            pool: r.pubkey(),
            admin: r.pubkey(),
            cancelledNewAdmin: r.pubkey(),
            slot: r.u64(),
          },
        };
      default:
        return null;
    }
  } catch {
    return null;
  }
}

/**
 * Decode a single log line into a DecodedEvent. Returns `null` for any line
 * that is not an event (or whose tag is unknown).
 */
export function decodeEventLog(rawLine: string): DecodedEvent | null {
  let line = rawLine;
  if (line.startsWith("Program log: ")) line = line.slice("Program log: ".length);
  else if (line.startsWith("Program data: ")) line = line.slice("Program data: ".length);
  if (!line.startsWith("EVT:")) return null;
  const encoded = line.slice("EVT:".length).trim();
  const payload = base64Decode(encoded);
  if (!payload || payload.length < 1) return null;
  const tag = payload[0];
  return decodeBody(tag, payload.subarray(1));
}

// ----------------------------------------------------------------------------
// Convenience: walk an entire transaction's logs
// ----------------------------------------------------------------------------

export function parseEventsFromLogs(logs: string[]): DecodedEvent[] {
  const out: DecodedEvent[] = [];
  for (const l of logs) {
    const ev = decodeEventLog(l);
    if (ev) out.push(ev);
  }
  return out;
}

/**
 * Fetch a confirmed transaction and parse all events from its logs. Accepts
 * either a connection or an AnchorProvider (back-compat with Anchor-era).
 */
export async function parseEventsFromTx(
  source: Connection | AnchorProvider,
  signature: string
): Promise<DecodedEvent[]> {
  const connection: Connection =
    "connection" in source ? (source as AnchorProvider).connection : source;
  await connection.confirmTransaction(signature, "confirmed");
  const tx = await connection.getTransaction(signature, {
    commitment: "confirmed",
    maxSupportedTransactionVersion: 0,
  });
  if (!tx?.meta?.logMessages) return [];
  return parseEventsFromLogs(tx.meta.logMessages);
}

/** Subscribe to a specific event type. Returns the listener id. */
export function subscribeEvent<TName extends ProtocolEventName>(
  target: {
    addEventListener: (
      name: TName,
      cb: (data: unknown, slot: number, sig: string) => void
    ) => number;
    removeEventListener: (id: number) => Promise<void>;
  },
  name: TName,
  callback: (
    data: ProtocolEventDataMap[TName],
    slot: number,
    signature: string
  ) => void
): number {
  return target.addEventListener(name, (data, slot, sig) =>
    callback(data as ProtocolEventDataMap[TName], slot, sig)
  );
}

export async function unsubscribeEvent(
  target: { removeEventListener: (id: number) => Promise<void> },
  listenerId: number
): Promise<void> {
  await target.removeEventListener(listenerId);
}

export { PROGRAM_ID };

File: programs/protocol/src/events.rs
use borsh::{BorshDeserialize, BorshSerialize};
use pinocchio::Address;

// ============================================================================
// Events
// ============================================================================
// Anchor's `emit!` macro Borsh-encoded the body and emitted
// `Program data: <base64>` log lines via `sol_log_data`. We keep the same
// shape (Borsh-encoded body, base64-wrapped) but emit via `sol_log_` instead,
// because the body is `[1-byte tag][Borsh body]` — the SDK can dispatch on
// the tag before deciding which struct to deserialize, without needing an IDL.
//
// Log format (one line per event):
//   Program log: EVT:<base64(tag || borsh_body)>
//
// `Program log: ` is the Solana runtime's prefix for the `sol_log_` syscall;
// `EVT:` is a literal identifier so callers can grep these lines without
// false-matching against unrelated "Program log:" entries. (The SDK decoder
// also accepts `Program data: EVT:…` lines for forward compatibility if we
// ever switch to `sol_log_data`.)
//
// Spec: docs/SPECIFICATION.md §3.

/// Stable event tag byte. Mirrored by the SDK event decoder.
pub mod tag {
    pub const POOL_INITIALIZED: u8 = 0x01;
    pub const ORACLE_UPDATED: u8 = 0x02;
    pub const SWAP_EXECUTED: u8 = 0x03;
    pub const POOL_PAUSED_CHANGED: u8 = 0x04;
    pub const ORACLE_SIGNER_ROTATED: u8 = 0x05;
    pub const ADMIN_ROTATED: u8 = 0x06;
    pub const INVENTORY_WITHDRAWN: u8 = 0x07;
    pub const QUOTE_MARKER_CLOSED: u8 = 0x08;
    pub const ADMIN_PROPOSAL_CREATED: u8 = 0x09;
    pub const ADMIN_PROPOSAL_CANCELLED: u8 = 0x0A;
    pub const QUOTE_SIGNER_ROTATED: u8 = 0x0B;
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct PoolInitialized {
    pub pool: Address,
    pub admin: Address,
    pub oracle_signer: Address,
    /// Initial RFQ quote signer (`pool.authorized_quote_signer`).
    pub quote_signer: Address,
    pub base_mint: Address,
    pub quote_mint: Address,
    pub initial_fair_value: u64,
    pub initial_spread_bps: u16,
    pub initial_mode_ttl: u8,
    pub slot: u64,
}

/// Hot-path event — the keeper emits this every 200 ms in Mode A. Trimmed
/// to the **minimum non-derivable fields**:
///   - `pool`            — kept; required to disambiguate when an indexer
///                         is subscribed to all program events.
///   - `oracle_signer`   — DROPPED; recoverable from `pool.authorized_oracle_signer`.
///   - `slot`            — DROPPED; available on `tx.slot` in tx metadata.
/// Saved ~37 bytes of Borsh body → ~400 CU/emit. See changelog v0.6.
#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct OracleUpdated {
    pub pool: Address,
    pub new_fair_value: u64,
    pub new_spread_bps: u16,
    pub new_nonce: u64,
    pub new_ttl: u8,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct SwapExecuted {
    pub pool: Address,
    pub user: Address,
    /// 0=Buy, 1=Sell
    pub direction: u8,
    /// 0=curve, 1=rfq
    pub mode: u8,
    pub input_amount: u64,
    pub output_amount: u64,
    pub execution_price: u64,
    pub quote_nonce: u64,
    pub slot: u64,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct PoolPausedChanged {
    pub pool: Address,
    pub admin: Address,
    pub paused: u8,
    pub slot: u64,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct OracleSignerRotated {
    pub pool: Address,
    pub admin: Address,
    pub previous_signer: Address,
    pub new_signer: Address,
    pub slot: u64,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct AdminRotated {
    pub pool: Address,
    pub previous_admin: Address,
    pub new_admin: Address,
    pub slot: u64,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct InventoryWithdrawn {
    pub pool: Address,
    pub admin: Address,
    pub base_amount: u64,
    pub quote_amount: u64,
    pub slot: u64,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct QuoteMarkerClosed {
    pub pool: Address,
    pub closer: Address,
    pub nonce: u64,
    pub expiry_slot: u64,
    pub slot: u64,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct AdminProposalCreated {
    pub pool: Address,
    pub proposed_by: Address,
    pub new_admin: Address,
    pub slot: u64,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct AdminProposalCancelled {
    pub pool: Address,
    pub admin: Address,
    pub cancelled_new_admin: Address,
    pub slot: u64,
}

#[derive(BorshSerialize, BorshDeserialize, Clone, Debug)]
pub struct QuoteSignerRotated {
    pub pool: Address,
    pub admin: Address,
    pub previous_signer: Address,
    pub new_signer: Address,
    pub slot: u64,
}

// ============================================================================
// Emit helpers
// ============================================================================
// Each emit_<name> packs `[tag | borsh-equivalent body bytes]` into a stack
// buffer and pushes it via `sol_log_data` — the runtime base64-encodes
// inside the syscall and emits `Program data: <base64>`. Byte layout per
// event is identical to what `BorshSerialize` would have produced (Borsh
// on a pure-fixed-size struct = field-order concatenation, no tags), so SDK
// decoders that read by offset keep working unchanged.
//
// CU history (per emit):
//   ~600–1500 CU  (v0.4 — sol_log_ + client-side base64)
//      ~300 CU    (v0.5 — sol_log_data + heap Vec)
//      ~150 CU    (v0.6 — sol_log_data + stack buffer)
//      ~100 CU    (v0.7 — sol_log_data + stack buffer + bytewise pack, no Borsh
//                  trait dispatch). Savings stack across every state-changing ix.

// Largest event body (`PoolInitialized` = 211 bytes) + 1 tag byte ≤ 256.
// Stack-allocated; no heap allocator path.
const EMIT_BUF_CAP: usize = 256;

/// Wrapper for `sol_log_data` with a single byte slice. Pinocchio 0.11 does
/// not re-export a safe wrapper, so we construct the fat-pointer-array layout
/// ourselves: `&[&[u8]] as *const u8` + the slice count as `len`.
#[inline(always)]
fn log_data_one(_bytes: &[u8]) {
    #[cfg(target_os = "solana")]
    unsafe {
        let slices: [&[u8]; 1] = [_bytes];
        pinocchio::syscalls::sol_log_data(
            slices.as_ptr() as *const u8,
            slices.len() as u64,
        );
    }
}

// ----- Per-event packers --------------------------------------------------
//
// Each event has a fixed byte size = sum of its field byte widths. Layout
// follows the struct declaration order so it stays byte-identical with what
// `BorshSerialize` would have produced — SDK decoders by offset are
// unchanged.

const SZ_POOL_INITIALIZED: usize = 6 * 32 + 8 + 2 + 1 + 8; // 211
const SZ_ORACLE_UPDATED: usize = 32 + 8 + 2 + 8 + 1;        // 51
const SZ_SWAP_EXECUTED: usize = 2 * 32 + 1 + 1 + 5 * 8;     // 106
const SZ_POOL_PAUSED_CHANGED: usize = 2 * 32 + 1 + 8;       // 73
const SZ_ORACLE_SIGNER_ROTATED: usize = 4 * 32 + 8;         // 136
const SZ_ADMIN_ROTATED: usize = 3 * 32 + 8;                 // 104
const SZ_INVENTORY_WITHDRAWN: usize = 2 * 32 + 8 + 8 + 8;   // 88
const SZ_QUOTE_MARKER_CLOSED: usize = 2 * 32 + 8 + 8 + 8;   // 88
const SZ_ADMIN_PROPOSAL_CREATED: usize = 3 * 32 + 8;        // 104
const SZ_ADMIN_PROPOSAL_CANCELLED: usize = 3 * 32 + 8;      // 104
const SZ_QUOTE_SIGNER_ROTATED: usize = 4 * 32 + 8;          // 136

pub fn emit_pool_initialized(e: &PoolInitialized) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::POOL_INITIALIZED;
    let b = &mut buf[1..1 + SZ_POOL_INITIALIZED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.admin.as_ref());
    b[64..96].copy_from_slice(e.oracle_signer.as_ref());
    b[96..128].copy_from_slice(e.quote_signer.as_ref());
    b[128..160].copy_from_slice(e.base_mint.as_ref());
    b[160..192].copy_from_slice(e.quote_mint.as_ref());
    b[192..200].copy_from_slice(&e.initial_fair_value.to_le_bytes());
    b[200..202].copy_from_slice(&e.initial_spread_bps.to_le_bytes());
    b[202] = e.initial_mode_ttl;
    b[203..211].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_POOL_INITIALIZED]);
}

/// Retained for legacy callers / completeness. NOT emitted by `update_oracle`
/// in v0.6+ — see SPEC §3.13 note.
pub fn emit_oracle_updated(e: &OracleUpdated) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::ORACLE_UPDATED;
    let b = &mut buf[1..1 + SZ_ORACLE_UPDATED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..40].copy_from_slice(&e.new_fair_value.to_le_bytes());
    b[40..42].copy_from_slice(&e.new_spread_bps.to_le_bytes());
    b[42..50].copy_from_slice(&e.new_nonce.to_le_bytes());
    b[50] = e.new_ttl;
    log_data_one(&buf[..1 + SZ_ORACLE_UPDATED]);
}

pub fn emit_swap_executed(e: &SwapExecuted) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::SWAP_EXECUTED;
    let b = &mut buf[1..1 + SZ_SWAP_EXECUTED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.user.as_ref());
    b[64] = e.direction;
    b[65] = e.mode;
    b[66..74].copy_from_slice(&e.input_amount.to_le_bytes());
    b[74..82].copy_from_slice(&e.output_amount.to_le_bytes());
    b[82..90].copy_from_slice(&e.execution_price.to_le_bytes());
    b[90..98].copy_from_slice(&e.quote_nonce.to_le_bytes());
    b[98..106].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_SWAP_EXECUTED]);
}

pub fn emit_pool_paused_changed(e: &PoolPausedChanged) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::POOL_PAUSED_CHANGED;
    let b = &mut buf[1..1 + SZ_POOL_PAUSED_CHANGED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.admin.as_ref());
    b[64] = e.paused;
    b[65..73].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_POOL_PAUSED_CHANGED]);
}

pub fn emit_oracle_signer_rotated(e: &OracleSignerRotated) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::ORACLE_SIGNER_ROTATED;
    let b = &mut buf[1..1 + SZ_ORACLE_SIGNER_ROTATED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.admin.as_ref());
    b[64..96].copy_from_slice(e.previous_signer.as_ref());
    b[96..128].copy_from_slice(e.new_signer.as_ref());
    b[128..136].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_ORACLE_SIGNER_ROTATED]);
}

pub fn emit_admin_rotated(e: &AdminRotated) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::ADMIN_ROTATED;
    let b = &mut buf[1..1 + SZ_ADMIN_ROTATED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.previous_admin.as_ref());
    b[64..96].copy_from_slice(e.new_admin.as_ref());
    b[96..104].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_ADMIN_ROTATED]);
}

pub fn emit_inventory_withdrawn(e: &InventoryWithdrawn) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::INVENTORY_WITHDRAWN;
    let b = &mut buf[1..1 + SZ_INVENTORY_WITHDRAWN];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.admin.as_ref());
    b[64..72].copy_from_slice(&e.base_amount.to_le_bytes());
    b[72..80].copy_from_slice(&e.quote_amount.to_le_bytes());
    b[80..88].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_INVENTORY_WITHDRAWN]);
}

pub fn emit_quote_marker_closed(e: &QuoteMarkerClosed) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::QUOTE_MARKER_CLOSED;
    let b = &mut buf[1..1 + SZ_QUOTE_MARKER_CLOSED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.closer.as_ref());
    b[64..72].copy_from_slice(&e.nonce.to_le_bytes());
    b[72..80].copy_from_slice(&e.expiry_slot.to_le_bytes());
    b[80..88].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_QUOTE_MARKER_CLOSED]);
}

pub fn emit_admin_proposal_created(e: &AdminProposalCreated) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::ADMIN_PROPOSAL_CREATED;
    let b = &mut buf[1..1 + SZ_ADMIN_PROPOSAL_CREATED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.proposed_by.as_ref());
    b[64..96].copy_from_slice(e.new_admin.as_ref());
    b[96..104].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_ADMIN_PROPOSAL_CREATED]);
}

pub fn emit_admin_proposal_cancelled(e: &AdminProposalCancelled) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::ADMIN_PROPOSAL_CANCELLED;
    let b = &mut buf[1..1 + SZ_ADMIN_PROPOSAL_CANCELLED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.admin.as_ref());
    b[64..96].copy_from_slice(e.cancelled_new_admin.as_ref());
    b[96..104].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_ADMIN_PROPOSAL_CANCELLED]);
}

pub fn emit_quote_signer_rotated(e: &QuoteSignerRotated) {
    let mut buf = [0u8; EMIT_BUF_CAP];
    buf[0] = tag::QUOTE_SIGNER_ROTATED;
    let b = &mut buf[1..1 + SZ_QUOTE_SIGNER_ROTATED];
    b[0..32].copy_from_slice(e.pool.as_ref());
    b[32..64].copy_from_slice(e.admin.as_ref());
    b[64..96].copy_from_slice(e.previous_signer.as_ref());
    b[96..128].copy_from_slice(e.new_signer.as_ref());
    b[128..136].copy_from_slice(&e.slot.to_le_bytes());
    log_data_one(&buf[..1 + SZ_QUOTE_SIGNER_ROTATED]);
}

#[cfg(test)]
mod size_tests {
    //! Compile-time guard: bytewise packers must produce the same byte count
    //! as Borsh serialize. If a struct field changes, both sides must update.

    use super::*;
    use borsh::BorshSerialize;

    fn borsh_len<T: BorshSerialize>(v: &T) -> usize {
        let mut buf = alloc::vec::Vec::new();
        v.serialize(&mut buf).unwrap();
        buf.len()
    }

    fn z_addr() -> Address {
        Address::default()
    }

    #[test]
    fn event_sizes_match_borsh() {
        assert_eq!(SZ_POOL_INITIALIZED, borsh_len(&PoolInitialized {
            pool: z_addr(), admin: z_addr(), oracle_signer: z_addr(),
            quote_signer: z_addr(), base_mint: z_addr(), quote_mint: z_addr(),
            initial_fair_value: 0, initial_spread_bps: 0, initial_mode_ttl: 0, slot: 0,
        }));
        assert_eq!(SZ_ORACLE_UPDATED, borsh_len(&OracleUpdated {
            pool: z_addr(), new_fair_value: 0, new_spread_bps: 0, new_nonce: 0, new_ttl: 0,
        }));
        assert_eq!(SZ_SWAP_EXECUTED, borsh_len(&SwapExecuted {
            pool: z_addr(), user: z_addr(), direction: 0, mode: 0,
            input_amount: 0, output_amount: 0, execution_price: 0, quote_nonce: 0, slot: 0,
        }));
        assert_eq!(SZ_POOL_PAUSED_CHANGED, borsh_len(&PoolPausedChanged {
            pool: z_addr(), admin: z_addr(), paused: 0, slot: 0,
        }));
        assert_eq!(SZ_ORACLE_SIGNER_ROTATED, borsh_len(&OracleSignerRotated {
            pool: z_addr(), admin: z_addr(), previous_signer: z_addr(),
            new_signer: z_addr(), slot: 0,
        }));
        assert_eq!(SZ_ADMIN_ROTATED, borsh_len(&AdminRotated {
            pool: z_addr(), previous_admin: z_addr(), new_admin: z_addr(), slot: 0,
        }));
        assert_eq!(SZ_INVENTORY_WITHDRAWN, borsh_len(&InventoryWithdrawn {
            pool: z_addr(), admin: z_addr(), base_amount: 0, quote_amount: 0, slot: 0,
        }));
        assert_eq!(SZ_QUOTE_MARKER_CLOSED, borsh_len(&QuoteMarkerClosed {
            pool: z_addr(), closer: z_addr(), nonce: 0, expiry_slot: 0, slot: 0,
        }));
        assert_eq!(SZ_ADMIN_PROPOSAL_CREATED, borsh_len(&AdminProposalCreated {
            pool: z_addr(), proposed_by: z_addr(), new_admin: z_addr(), slot: 0,
        }));
        assert_eq!(SZ_ADMIN_PROPOSAL_CANCELLED, borsh_len(&AdminProposalCancelled {
            pool: z_addr(), admin: z_addr(), cancelled_new_admin: z_addr(), slot: 0,
        }));
        assert_eq!(SZ_QUOTE_SIGNER_ROTATED, borsh_len(&QuoteSignerRotated {
            pool: z_addr(), admin: z_addr(), previous_signer: z_addr(),
            new_signer: z_addr(), slot: 0,
        }));
    }
}
